### PR TITLE
Future public types for @ember/controller

### PR DESCRIPTION
**Pull request by wagenet**
Includes #19939

### CODE REVIEW

File: packages/@ember/-internals/metal/lib/injected_property.ts [hunk 1]
```diff
@@ -62,12 +81,12 @@ function inject(type: string, ...args: any[]): Decorator | DecoratorPropertyDesc
...
     get: getInjection,
 
-    set(this: any, keyName: string, value: any) {
+    set(this: object, keyName: string, value: unknown) {
       defineProperty(this, keyName, null, value);
     },
   });
 
-  if (calledAsDecorator) {
+  if (isElementDescriptor(args)) {
```
Comment (wagenet) [inline packages/@ember/-internals/metal/lib/injected_property.ts:89]: Having to call this twice is annoying, but probably worth it for the increased typing benefits. This isn't a hot path.

File: packages/@ember/-internals/routing/lib/system/router.ts [hunk 2]
```diff
@@ -706,7 +712,7 @@ class EmberRouter extends EmberObject.extend(Evented) implements Evented {
     }
   }
 
-  replaceWith(...args: any[]) {
+  replaceWith(...args: RouteArgs): Transition {
```
Comment (wagenet) [inline packages/@ember/-internals/routing/lib/system/router.ts:715]: Should this have docs?

File: packages/@ember/controller/index.ts [hunk 3]
```diff
@@ -55,8 +63,19 @@ const Controller = FrameworkObject.extend(ControllerMixin);
...
+  target: ElementDescriptor[0],
+  key: ElementDescriptor[1],
+  desc: ElementDescriptor[2]
+): DecoratorPropertyDescriptor;
+export function inject(...args: unknown[]): Decorator | DecoratorPropertyDescriptor {
+  if (isElementDescriptor(args)) {
+    return metalInject('controller', ...args);
+  } else {
+    assert('name is string or undefined', args[0] === undefined || typeof args[0] === 'string');
+    return metalInject('controller', args[0]);
```
Comment (wagenet) [inline packages/@ember/controller/index.ts:77]: This method body is unfortunate. Alternatives welcome!

File: packages/@ember/-internals/routing/lib/system/router.ts (hunk 2 above)
Comment (chriskrycho) [reply packages/@ember/-internals/routing/lib/system/router.ts:715]: I think *yes and they should mark it as deprecated, same as `transitionTo`*.

**Review by chriskrycho — CHANGES_REQUESTED** [review]
WIP review, more to come tomorrow, but wanted to provide the feedback I have so far.

File: packages/@ember/-internals/metal/lib/injected_property.ts [hunk 4]
```diff
@@ -27,26 +36,36 @@ if (DEBUG) {
...
-  key: string,
-  desc: DecoratorPropertyDescriptor
+  target: ElementDescriptor[0],
+  key: ElementDescriptor[1],
+  desc: ElementDescriptor[2]
 ): DecoratorPropertyDescriptor;
-function inject(type: string, ...args: any[]): Decorator | DecoratorPropertyDescriptor {
+function inject(
+  type: string,
+  ...args: [name: string | undefined] | ElementDescriptor
```
[the remark this answers is not in the document]
Comment (wagenet) [reply packages/@ember/-internals/metal/lib/injected_property.ts:48]: @chriskrycho `ElementDescriptor` is a tuple type.

File: packages/@ember/-internals/metal/lib/injected_property.ts [hunk 5]
```diff
@@ -27,26 +36,36 @@ if (DEBUG) {
...
+): Decorator | DecoratorPropertyDescriptor {
   assert('a string type must be provided to inject', typeof type === 'string');
 
-  let calledAsDecorator = isElementDescriptor(args);
+  let name: string | undefined;
+
+  if (!isElementDescriptor(args)) {
```
[the remark this answers is not in the document]
Comment (wagenet) [reply packages/@ember/-internals/metal/lib/injected_property.ts:54]: Hmm... that wasn't working correctly for me, but I could try a bit more.

File: packages/@ember/-internals/metal/lib/injected_property.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (wagenet) [reply packages/@ember/-internals/metal/lib/injected_property.ts:48]: https://github.com/emberjs/ember.js/blob/master/packages/@ember/-internals/metal/lib/decorator.ts#L16

File: packages/@ember/-internals/metal/lib/injected_property.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (wagenet) [reply packages/@ember/-internals/metal/lib/injected_property.ts:54]: I just tried reverting and it does not correctly determine the type.

File: packages/@ember/-internals/metal/lib/injected_property.ts (hunk 1 above)
Comment (wagenet) [reply packages/@ember/-internals/metal/lib/injected_property.ts:89]: Yeah, doesn't work on 4.5.

File: packages/@ember/-internals/routing/lib/system/router.ts (hunk 2 above)
Comment (wagenet) [reply packages/@ember/-internals/routing/lib/system/router.ts:715]: `Router#transitionTo` isn't marked as deprecated.

File: packages/@ember/-internals/metal/lib/injected_property.ts [hunk 6]
```diff
@@ -27,26 +36,36 @@ if (DEBUG) {
...
-  desc: DecoratorPropertyDescriptor
+  target: ElementDescriptor[0],
+  key: ElementDescriptor[1],
+  desc: ElementDescriptor[2]
 ): DecoratorPropertyDescriptor;
-function inject(type: string, ...args: any[]): Decorator | DecoratorPropertyDescriptor {
+function inject(
+  type: string,
+  ...args: [name: string | undefined] | ElementDescriptor
+): Decorator | DecoratorPropertyDescriptor {
```
Comment (wagenet) [inline packages/@ember/-internals/metal/lib/injected_property.ts:49]: There's also a question of whether we want to introduce a type Registry here.

File: packages/@ember/-internals/metal/lib/injected_property.ts [hunk 7]
```diff
@@ -27,26 +36,36 @@ if (DEBUG) {
...
+function inject(type: string, name: string | undefined): Decorator;
 function inject(
   type: string,
-  target: object,
-  key: string,
-  desc: DecoratorPropertyDescriptor
+  target: ElementDescriptor[0],
+  key: ElementDescriptor[1],
+  desc: ElementDescriptor[2]
 ): DecoratorPropertyDescriptor;
```
Comment (wagenet) [inline packages/@ember/-internals/metal/lib/injected_property.ts:45]: We need to make sure this works in the decorator form.

File: packages/@ember/-internals/metal/lib/injected_property.ts (hunk 1 above)
Comment (wagenet) [reply packages/@ember/-internals/metal/lib/injected_property.ts]: I refactored this.

File: packages/@ember/controller/index.ts (hunk 3 above)
Comment (wagenet) [reply packages/@ember/controller/index.ts]: I refactored this.

File: packages/@ember/-internals/metal/lib/injected_property.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (wagenet) [reply packages/@ember/-internals/metal/lib/injected_property.ts]: I refactored this.

File: packages/@ember/-internals/metal/lib/injected_property.ts (hunk 7 above)
Comment (wagenet) [reply packages/@ember/-internals/metal/lib/injected_property.ts]: I refactored this.

**Pull request merged**